From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mortyzhaoyu <morty.zhaoyu@gmail.com>
Date: Mon, 23 Jun 2026 12:00:00 +0800
Subject: [PATCH] rk3576-remotectl

--- a/drivers/input/Kconfig	2026-05-10 23:20:52.000000000 -0700
+++ b/drivers/input/Kconfig	2026-06-23 04:10:07.653979472 -0700
@@ -188,6 +188,8 @@
 
 source "drivers/input/misc/Kconfig"
 
+source "drivers/input/remotectl/Kconfig"
+
 source "drivers/input/rmi4/Kconfig"
 
 endif
--- a/drivers/input/Makefile	2026-05-10 23:20:52.000000000 -0700
+++ b/drivers/input/Makefile	2026-06-23 04:10:26.919597897 -0700
@@ -25,6 +25,7 @@
 obj-$(CONFIG_INPUT_TABLET)	+= tablet/
 obj-$(CONFIG_INPUT_TOUCHSCREEN)	+= touchscreen/
 obj-$(CONFIG_INPUT_MISC)	+= misc/
+obj-$(CONFIG_ROCKCHIP_REMOTECTL)	+= remotectl/
 obj-$(CONFIG_INPUT_KUNIT_TEST)	+= tests/
 
 obj-$(CONFIG_INPUT_APMPOWER)	+= apm-power.o
diff -ruN a/drivers/input/remotectl/Kconfig b/drivers/input/remotectl/Kconfig
--- a/drivers/input/remotectl/Kconfig	1969-12-31 16:00:00.000000000 -0800
+++ b/drivers/input/remotectl/Kconfig	2026-06-23 04:06:29.910977511 -0700
@@ -0,0 +1,19 @@
+# SPDX-License-Identifier: GPL-2.0
+#
+# Input remotectl driver configuration
+#
+menuconfig ROCKCHIP_REMOTECTL
+	tristate "rockchip remotectl"
+	default n
+	help
+	  Say Y here, will suport rk remotectl.
+	  This option doesn't affect the kernel.
+	  If unsure, say Y.
+
+if ROCKCHIP_REMOTECTL
+
+config ROCKCHIP_REMOTECTL_PWM
+	tristate "rockchip remoctrl pwm capture"
+	default n
+
+endif
diff -ruN a/drivers/input/remotectl/Makefile b/drivers/input/remotectl/Makefile
--- a/drivers/input/remotectl/Makefile	1969-12-31 16:00:00.000000000 -0800
+++ b/drivers/input/remotectl/Makefile	2026-06-23 04:06:29.912977575 -0700
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: GPL-2.0
+#
+# Makefile for the input remotectl drivers.
+#
+
+# Each configuration option enables a list of files.
+
+obj-$(CONFIG_ROCKCHIP_REMOTECTL_PWM)	+= rockchip_pwm_remotectl.o
diff -ruN a/drivers/input/remotectl/rockchip_pwm_remotectl.c b/drivers/input/remotectl/rockchip_pwm_remotectl.c
--- a/drivers/input/remotectl/rockchip_pwm_remotectl.c	1969-12-31 16:00:00.000000000 -0800
+++ b/drivers/input/remotectl/rockchip_pwm_remotectl.c	2026-06-23 04:10:02.233805422 -0700
@@ -0,0 +1,1087 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#include <linux/clk.h>
+#include <linux/io.h>
+#include <linux/irq.h>
+#include <linux/irqdesc.h>
+#include <linux/irqdomain.h>
+#include <linux/module.h>
+#include <linux/of.h>
+#include <linux/of_device.h>
+#include <linux/platform_device.h>
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/interrupt.h>
+#include <linux/platform_device.h>
+#include <linux/input.h>
+#include <linux/workqueue.h>
+#include <linux/slab.h>
+#include <linux/rockchip/rockchip_sip.h>
+#include "rockchip_pwm_remotectl.h"
+
+/*
+ * sys/module/rk_pwm_remotectl/parameters,
+ * modify code_print to change the value
+ */
+
+static int rk_remote_print_code;
+module_param_named(code_print, rk_remote_print_code, int, 0644);
+#define DBG_CODE(args...) \
+	do { \
+		if (rk_remote_print_code) { \
+			pr_info(args); \
+		} \
+	} while (0)
+
+static int rk_remote_pwm_dbg_level;
+module_param_named(dbg_level, rk_remote_pwm_dbg_level, int, 0644);
+#define DBG(args...) \
+	do { \
+		if (rk_remote_pwm_dbg_level) { \
+			pr_info(args); \
+		} \
+	} while (0)
+
+
+struct rkxx_remote_key_table {
+	int scancode;
+	int keycode;
+};
+
+struct rkxx_remotectl_button {
+	int usercode;
+	int nbuttons;
+	struct rkxx_remote_key_table key_table[MAX_NUM_KEYS];
+};
+
+struct rk_remote_pwm_regs {
+	unsigned long ctrl;
+	unsigned long version;
+	unsigned long enable;
+	unsigned long clk_ctrl;
+	unsigned long offset;
+	unsigned long rpt;
+	unsigned long hpr;
+	unsigned long lpr;
+	unsigned long intsts;
+	unsigned long int_en;
+	unsigned long pwrmatch_arbiter;
+	unsigned long pwrmatch_ctrl;
+	unsigned long pwrmatch_lpre;
+	unsigned long pwrmatch_hpre;
+	unsigned long pwrmatch_ld;
+	unsigned long pwrmatch_hd_zero;
+	unsigned long pwrmatch_hd_one;
+	unsigned long pwrmatch_value0;
+	unsigned long pwrcapture_value;
+};
+
+struct rk_remote_pwm_funcs {
+	irqreturn_t (*irq_handler)(int irq, void *data);
+	void (*int_ctrl)(struct platform_device *pdev, int work_mode);
+	int (*init_hw)(struct platform_device *pdev);
+};
+
+struct rk_remote_pwm_data {
+	struct rk_remote_pwm_regs regs;
+	struct rk_remote_pwm_funcs funcs;
+	unsigned int prescaler;
+	u32 enable_conf;
+	int pwm_version;
+	int pwrmactch_key_max;
+};
+
+struct rkxx_remotectl_drvdata {
+	void __iomem *base;
+	int pwm_version;
+	int state;
+	int nbuttons;
+	int scandata;
+	int count;
+	int keynum;
+	int maxkeybdnum;
+	int keycode;
+	int press;
+	int pre_press;
+	int irq;
+	int remote_pwm_id;
+	int handle_cpu_id;
+	int wakeup;
+	int support_psci;
+	int pwm_pwrkey_capture;
+	unsigned long period;
+	unsigned long temp_period;
+	int pwm_freq_nstime;
+	int pwrkey_wakeup;
+	struct input_dev *input;
+	struct device *dev;
+	struct timer_list timer;
+	struct tasklet_struct remote_tasklet;
+	const struct rk_remote_pwm_data *pwm_data;
+};
+
+static irqreturn_t rockchip_pwm_irq(int irq, void *dev_id);
+static irqreturn_t rockchip_pwm_irq_v4(int irq, void *dev_id);
+static int rk_pwm_remotectl_hw_init(struct platform_device *pdev);
+static int rk_pwm_remotectl_hw_init_v4(struct platform_device *pdev);
+static void rk_pwm_int_ctrl(struct platform_device *pdev, int work_mode);
+static void rk_pwm_int_ctrl_v4(struct platform_device *pdev, int work_mode);
+
+static struct rkxx_remotectl_button *remotectl_button;
+
+struct rk_remote_pwm_data pwm_data_v1 = {
+	.regs = {
+		.version = 0x5c,
+		.ctrl = 0x0c,
+	},
+	.prescaler = 1,
+	.pwm_version = 1,
+	.funcs = {
+		.irq_handler = rockchip_pwm_irq,
+		.int_ctrl = rk_pwm_int_ctrl,
+		.init_hw = rk_pwm_remotectl_hw_init,
+	},
+};
+
+struct rk_remote_pwm_data pwm_data_v4 = {
+	.regs = {
+		.version = 0x0,
+		.enable = 0x4,
+		.clk_ctrl = 0x8,
+		.ctrl = 0xc,
+		.hpr = 0x2c,
+		.lpr = 0x30,
+		.intsts = 0x70,
+		.int_en = 0x74,
+		.pwrmatch_arbiter = 0x100,
+		.pwrmatch_ctrl = 0x104,
+		.pwrmatch_lpre = 0x108,
+		.pwrmatch_hpre = 0x10c,
+		.pwrmatch_ld = 0x110,
+		.pwrmatch_hd_zero = 0x114,
+		.pwrmatch_hd_one = 0x118,
+		.pwrmatch_value0 = 0x11c,
+		.pwrcapture_value = 0x15c,
+	},
+	.pwm_version = 4,
+	.funcs = {
+		.irq_handler = rockchip_pwm_irq_v4,
+		.int_ctrl = rk_pwm_int_ctrl_v4,
+		.init_hw = rk_pwm_remotectl_hw_init_v4,
+	},
+};
+
+static int remotectl_keybd_num_lookup(struct rkxx_remotectl_drvdata *ddata)
+{
+	int i;
+	int num;
+
+	num = ddata->maxkeybdnum;
+	for (i = 0; i < num; i++) {
+		if (remotectl_button[i].usercode == (ddata->scandata&0xFFFF)) {
+			ddata->keynum = i;
+			return 1;
+		}
+	}
+	return 0;
+}
+
+static int remotectl_keycode_lookup(struct rkxx_remotectl_drvdata *ddata)
+{
+	int i;
+	unsigned char keydata = (unsigned char)((ddata->scandata >> 8) & 0xff);
+
+	for (i = 0; i < remotectl_button[ddata->keynum].nbuttons; i++) {
+		if (remotectl_button[ddata->keynum].key_table[i].scancode ==
+		    keydata) {
+			ddata->keycode =
+			remotectl_button[ddata->keynum].key_table[i].keycode;
+			return 1;
+		}
+	}
+	return 0;
+}
+
+static int rk_remotectl_get_irkeybd_count(struct platform_device *pdev)
+{
+	struct device_node *node = pdev->dev.of_node;
+	struct device_node *child_node;
+	int boardnum;
+	int temp_usercode;
+
+	boardnum = 0;
+	for_each_child_of_node(node, child_node) {
+		if (of_property_read_u32(child_node, "rockchip,usercode",
+			&temp_usercode)) {
+			DBG("get keybd num error.\n");
+		} else {
+			boardnum++;
+		}
+	}
+	DBG("get keybd num = 0x%x.\n", boardnum);
+	return boardnum;
+}
+
+static int rk_remotectl_parse_ir_keys(struct platform_device *pdev)
+{
+	struct rkxx_remotectl_drvdata *ddata = platform_get_drvdata(pdev);
+	struct device_node *node = pdev->dev.of_node;
+	struct device_node *child_node;
+	int loop;
+	int ret;
+	int len;
+	int boardnum;
+
+	boardnum = 0;
+	for_each_child_of_node(node, child_node) {
+		if (of_property_read_u32(child_node, "rockchip,usercode",
+			 &remotectl_button[boardnum].usercode)) {
+			dev_err(&pdev->dev, "Missing usercode in the DTS.\n");
+			ret = -1;
+			return ret;
+		}
+		DBG("remotectl_button[0].usercode=0x%x\n",
+				remotectl_button[boardnum].usercode);
+		of_get_property(child_node, "rockchip,key_table", &len);
+		len /= sizeof(u32);
+		DBG("len=0x%x\n", len);
+		remotectl_button[boardnum].nbuttons = len/2;
+		if (of_property_read_u32_array(child_node, "rockchip,key_table",
+			 (u32 *)remotectl_button[boardnum].key_table, len)) {
+			dev_err(&pdev->dev, "Missing key_table in the DTS.\n");
+			ret = -1;
+			return ret;
+		}
+		for (loop = 0; loop < (len/2); loop++) {
+			DBG("board[%d].scanCode[%d]=0x%x\n", boardnum, loop,
+			     remotectl_button[boardnum].key_table[loop].scancode);
+			DBG("board[%d].keyCode[%d]=%d\n", boardnum, loop,
+			     remotectl_button[boardnum].key_table[loop].keycode);
+		}
+		boardnum++;
+		if (boardnum > ddata->maxkeybdnum)
+			break;
+	}
+	DBG("keybdNum=0x%x\n", boardnum);
+	return 0;
+}
+
+static void rk_pwm_remotectl_do_something(unsigned long  data)
+{
+	struct rkxx_remotectl_drvdata *ddata;
+
+	ddata = (struct rkxx_remotectl_drvdata *)data;
+	switch (ddata->state) {
+	case RMC_IDLE: {
+		;
+		break;
+	}
+	case RMC_PRELOAD: {
+		mod_timer(&ddata->timer, jiffies + msecs_to_jiffies(140));
+		if ((ddata->period > RK_PWM_TIME_PRE_MIN) &&
+		    (ddata->period < RK_PWM_TIME_PRE_MAX)) {
+			ddata->scandata = 0;
+			ddata->count = 0;
+			ddata->state = RMC_USERCODE;
+		} else {
+			ddata->state = RMC_PRELOAD;
+		}
+		break;
+	}
+	case RMC_USERCODE: {
+		if ((ddata->period > RK_PWM_TIME_BIT1_MIN) &&
+		    (ddata->period < RK_PWM_TIME_BIT1_MAX))
+			ddata->scandata |= (0x01 << ddata->count);
+		ddata->count++;
+		if (ddata->count == 0x10) {
+			DBG_CODE("USERCODE=0x%x\n", ddata->scandata);
+			if (remotectl_keybd_num_lookup(ddata)) {
+				ddata->state = RMC_GETDATA;
+				ddata->scandata = 0;
+				ddata->count = 0;
+			} else {
+				if (rk_remote_print_code) {
+					ddata->state = RMC_GETDATA;
+					ddata->scandata = 0;
+					ddata->count = 0;
+				} else
+					ddata->state = RMC_PRELOAD;
+			}
+		}
+	}
+	break;
+	case RMC_GETDATA: {
+		if ((ddata->period > RK_PWM_TIME_BIT1_MIN) &&
+		    (ddata->period < RK_PWM_TIME_BIT1_MAX))
+			ddata->scandata |= (0x01<<ddata->count);
+		ddata->count++;
+		if (ddata->count < 0x10)
+			return;
+		DBG_CODE("RMC_GETDATA=%x\n", (ddata->scandata>>8));
+		if ((ddata->scandata&0x0ff) ==
+		    ((~ddata->scandata >> 8) & 0x0ff)) {
+			if (remotectl_keycode_lookup(ddata)) {
+				ddata->press = 1;
+				input_event(ddata->input, EV_KEY,
+					    ddata->keycode, 1);
+				input_sync(ddata->input);
+				ddata->state = RMC_SEQUENCE;
+			} else {
+				ddata->state = RMC_PRELOAD;
+			}
+		} else {
+			ddata->state = RMC_PRELOAD;
+		}
+	}
+	break;
+	case RMC_SEQUENCE:{
+		DBG("S=%ld\n", ddata->period);
+		if ((ddata->period > RK_PWM_TIME_RPT_MIN) &&
+		    (ddata->period < RK_PWM_TIME_RPT_MAX)) {
+			DBG("S1\n");
+			mod_timer(&ddata->timer, jiffies
+				  + msecs_to_jiffies(130));
+		} else if ((ddata->period > RK_PWM_TIME_SEQ1_MIN) &&
+			   (ddata->period < RK_PWM_TIME_SEQ1_MAX)) {
+			DBG("S2\n");
+			mod_timer(&ddata->timer, jiffies
+				  + msecs_to_jiffies(130));
+		} else if ((ddata->period > RK_PWM_TIME_SEQ2_MIN) &&
+			   (ddata->period < RK_PWM_TIME_SEQ2_MAX)) {
+			DBG("S3\n");
+			mod_timer(&ddata->timer, jiffies
+				  + msecs_to_jiffies(130));
+		} else {
+			DBG("S4\n");
+			input_event(ddata->input, EV_KEY,
+				    ddata->keycode, 0);
+			input_sync(ddata->input);
+			ddata->state = RMC_PRELOAD;
+			ddata->press = 0;
+		}
+	}
+	break;
+	default:
+	break;
+	}
+}
+
+static void rk_pwm_remotectl_timer(struct timer_list *t)
+{
+	struct rkxx_remotectl_drvdata *ddata = from_timer(ddata, t, timer);
+
+	if (ddata->press != ddata->pre_press) {
+		ddata->pre_press = 0;
+		ddata->press = 0;
+		input_event(ddata->input, EV_KEY, ddata->keycode, 0);
+		input_sync(ddata->input);
+	}
+	ddata->state = RMC_PRELOAD;
+}
+
+static irqreturn_t rockchip_pwm_pwrirq(int irq, void *dev_id)
+{
+	struct rkxx_remotectl_drvdata *ddata = dev_id;
+	int val;
+	unsigned int id = ddata->remote_pwm_id;
+
+	if (id > 3)
+		return IRQ_NONE;
+
+	val = readl_relaxed(ddata->base + PWM_REG_INTSTS(id));
+
+	if (val & PWM_PWR_KEY_INT) {
+		DBG("pwr=0x%x\n", readl_relaxed(ddata->base + PWM_PWRCAPTURE_VALUE(id)));
+		writel_relaxed(PWM_PWR_KEY_INT, ddata->base + PWM_REG_INTSTS(id));
+		ddata->pwrkey_wakeup = 1;
+		return IRQ_HANDLED;
+	}
+
+	return IRQ_NONE;
+}
+
+static irqreturn_t rockchip_pwm_irq(int irq, void *dev_id)
+{
+	struct rkxx_remotectl_drvdata *ddata = dev_id;
+	int val;
+	int temp_hpr;
+	int temp_lpr;
+	int temp_period;
+	unsigned int id = ddata->remote_pwm_id;
+
+	if (id > 3)
+		return IRQ_NONE;
+	val = readl_relaxed(ddata->base + PWM_REG_INTSTS(id));
+
+	if ((val & PWM_CH_INT(id)) == 0)
+		return IRQ_NONE;
+	if ((val & PWM_CH_POL(id)) == 0) {
+		temp_hpr = readl_relaxed(ddata->base + PWM_REG_HPR);
+		writel_relaxed(0, ddata->base + PWM_REG_HPR);
+		temp_lpr = readl_relaxed(ddata->base + PWM_REG_LPR);
+		writel_relaxed(0, ddata->base + PWM_REG_LPR);
+		DBG("hpr=%d\n", temp_hpr);
+		DBG("lpr=%d\n", temp_lpr);
+
+		temp_period = ddata->pwm_freq_nstime * temp_lpr / 1000;
+		if (temp_period > RK_PWM_TIME_BIT0_MIN) {
+			ddata->period = ddata->temp_period
+			    + ddata->pwm_freq_nstime * temp_hpr / 1000;
+			tasklet_hi_schedule(&ddata->remote_tasklet);
+			ddata->temp_period = 0;
+			DBG("period+ =%ld\n", ddata->period);
+		} else {
+			ddata->temp_period += ddata->pwm_freq_nstime
+			    * (temp_hpr + temp_lpr) / 1000;
+		}
+	}
+	writel_relaxed(PWM_CH_INT(id), ddata->base + PWM_REG_INTSTS(id));
+	if (ddata->state == RMC_PRELOAD)
+		pm_wakeup_dev_event(ddata->dev, 1000, true);
+	return IRQ_HANDLED;
+}
+
+static irqreturn_t rockchip_pwm_irq_v4(int irq, void *dev_id)
+{
+	struct rkxx_remotectl_drvdata *ddata = dev_id;
+	int tmp;
+	int val;
+	irqreturn_t ret = IRQ_NONE;
+
+	val = readl_relaxed(ddata->base + PWM_REG_INTSTS_V4);
+
+	if (val & CAP_LPR_INT) {
+		writel_relaxed(CAP_LPR_INT, ddata->base + PWM_REG_INTSTS_V4);
+		tmp = readl_relaxed(ddata->base + PWM_REG_HPR_V4);
+		ddata->period  = ddata->pwm_freq_nstime * tmp / 1000;
+		DBG("period = %ld\n", ddata->period);
+		tasklet_hi_schedule(&ddata->remote_tasklet);
+		ret = IRQ_HANDLED;
+	} else if (val & CAP_HPR_INT) {
+		writel_relaxed(CAP_HPR_INT, ddata->base + PWM_REG_INTSTS_V4);
+		tmp = readl_relaxed(ddata->base + PWM_REG_LPR_V4);
+		DBG("lpr = %d\n", tmp);
+		ret = IRQ_HANDLED;
+	}
+
+	if (val & PWR_INT) {
+		writel_relaxed(PWR_INT, ddata->base + PWM_REG_INTSTS_V4);
+		tmp = readl_relaxed(ddata->base + PWM_REG_CAPTURE_VALUE_V4);
+		DBG("##### cap_val = 0x%0x\n", tmp);
+		ddata->pwrkey_wakeup = 1;
+		ret = IRQ_HANDLED;
+	}
+
+	if (ddata->state == RMC_PRELOAD)
+		pm_wakeup_dev_event(ddata->dev, 1000, true);
+
+	return ret;
+}
+
+static int rk_pwmkey_convert_val(int min, int max, int freq_nstime)
+{
+	int val, min_temp, max_temp;
+
+	min_temp = min * 1000 / freq_nstime;
+	max_temp = max * 1000 / freq_nstime;
+	val = (max_temp & 0xffff) << 16 | (min_temp & 0xffff);
+
+	return val;
+}
+
+static int rockchip_pwm_set_pwrmatch(struct platform_device *pdev)
+{
+	struct rkxx_remotectl_drvdata *ddata = platform_get_drvdata(pdev);
+	unsigned int pwm_id = ddata->remote_pwm_id;
+	int version;
+	int pwr_irq;
+	int val;
+	int ret;
+
+	version = readl_relaxed(ddata->base + RK_PWM_VERSION_ID(pwm_id));
+	DBG("remote pwm version is 0x%x\n", version);
+	if (((version >> 24) & 0xFF) < 2) {
+		dev_info(&pdev->dev, "pwm version is less v2.0\n");
+		ret = -EINVAL;
+		goto end;
+	}
+	pwr_irq = platform_get_irq(pdev, 1);
+	if (pwr_irq < 0) {
+		dev_err(&pdev->dev, "cannot find PWR IRQ\n");
+		ret = pwr_irq;
+		goto end;
+	}
+	ret = devm_request_irq(&pdev->dev, pwr_irq, rockchip_pwm_pwrirq,
+			IRQF_NO_SUSPEND, "rk_pwm_pwr_irq", ddata);
+	if (ret) {
+		dev_err(&pdev->dev, "cannot claim PWR_IRQ!!!\n");
+		goto end;
+	}
+	val = readl_relaxed(ddata->base + PWM_REG_INT_EN(pwm_id));
+	val = (val & 0xFFFFFF7F) | PWM_PWR_INT_ENABLE;
+	writel_relaxed(val, ddata->base + PWM_REG_INT_EN(pwm_id));
+
+	val = CH3_PWRKEY_ENABLE;
+	writel_relaxed(val, ddata->base + PWM_REG_PWRMATCH_CTRL(pwm_id));
+
+	val = readl_relaxed(ddata->base + PWM_REG_CTRL);
+	val = (val & 0xFFFFFFFE) | PWM_ENABLE;
+	writel_relaxed(val, ddata->base + PWM_REG_CTRL);
+	return 0;
+end:
+	return ret;
+}
+
+static int rockchip_pwm_set_pwrmatch_v4(struct rkxx_remotectl_drvdata *pd)
+{
+	int version;
+	int channel_id;
+	int val;
+
+	version = readl_relaxed(pd->base + PWM_REG_VERSION_V4);
+	DBG("remote pwm version is 0x%x\n", version);
+	channel_id = (version & CHANNLE_INDEX_MASK) >> CHANNLE_INDEX_SHIFT;
+	val = BIT(channel_id) << PWRMATCH_READ_LOCK_SHIFT |
+			BIT(channel_id) << PWRMATCH_GRANT_SHIFT;
+	writel_relaxed(val, pd->base + PWM_REG_MATCH_ARBITER_V4);
+	writel_relaxed(PWRKEY_EN(true),
+		       pd->base + PWM_REG_MATCH_CTRL_V4);
+
+	return 0;
+}
+
+static void rockchip_pwrmatch_set_nec_param(struct rkxx_remotectl_drvdata *pd)
+{
+	unsigned int pwm_id = pd->remote_pwm_id;
+	void __iomem *reg;
+	int freq_nstime;
+	int val;
+
+	freq_nstime = pd->pwm_freq_nstime;
+	//preloader low min:8000us, max:10000us
+	val = rk_pwmkey_convert_val(RK_PWM_TIME_PRE_MIN_LOW,
+				    RK_PWM_TIME_PRE_MAX_LOW, freq_nstime);
+	if (pd->pwm_data->pwm_version < 4)
+		reg = pd->base + PWM_REG_PWRMATCH_LPRE(pwm_id);
+	else
+		reg = pd->base + PWM_REG_MATCH_LPRE_V4;
+	writel_relaxed(val, reg);
+
+	//preloader higt min:4000us, max:5000us
+	val = rk_pwmkey_convert_val(RK_PWM_TIME_PRE_MIN,
+				    RK_PWM_TIME_PRE_MAX, freq_nstime);
+	if (pd->pwm_data->pwm_version < 4)
+		reg = pd->base + PWM_REG_PWRMATCH_HPRE(pwm_id);
+	else
+		reg = pd->base + PWM_REG_MATCH_HPRE_V4;
+	writel_relaxed(val, reg);
+
+	//logic 0/1 low min:480us, max 700us
+	val = rk_pwmkey_convert_val(RK_PWM_TIME_BIT_MIN_LOW,
+				    RK_PWM_TIME_BIT_MAX_LOW, freq_nstime);
+	if (pd->pwm_data->pwm_version < 4)
+		reg = pd->base + PWM_REG_PWRMATCH_LD(pwm_id);
+	else
+		reg = pd->base + PWM_REG_MATCH_LD_V4;
+	writel_relaxed(val, reg);
+
+	//logic 0 higt min:480us, max 700us
+	val = rk_pwmkey_convert_val(RK_PWM_TIME_BIT0_MIN,
+				    RK_PWM_TIME_BIT0_MAX, freq_nstime);
+	if (pd->pwm_data->pwm_version < 4)
+		reg = pd->base + PWM_REG_PWRMATCH_HD_ZERO(pwm_id);
+	else
+		reg = pd->base + PWM_REG_MATCH_HD_ZERO_V4;
+	writel_relaxed(val, reg);
+
+	//logic 1 higt min:1300us, max 2000us
+	val = rk_pwmkey_convert_val(RK_PWM_TIME_BIT1_MIN,
+				    RK_PWM_TIME_BIT1_MAX, freq_nstime);
+	if (pd->pwm_data->pwm_version < 4)
+		reg = pd->base + PWM_REG_PWRMATCH_HD_ONE(pwm_id);
+	else
+		reg = pd->base + PWM_REG_MATCH_HD_ONE_V4;
+	writel_relaxed(val, reg);
+}
+
+static void rockchip_pwrmatch_set_pwrkey(struct rkxx_remotectl_drvdata *pd)
+{
+	unsigned int pwm_id = pd->remote_pwm_id;
+	void __iomem *reg;
+	int i, j;
+	int num = 0, num_max;
+
+	if (pd->pwm_data->pwm_version < 4) {
+		reg = pd->base + PWM_PWRMATCH_VALUE(pwm_id);
+		num_max = PWM_PWR_KEY_CAPURURE_MAX;
+	} else {
+		reg = pd->base + PWM_REG_CAPTURE_VALUE0_V4;
+		num_max = PWM_PWR_KEY_CAPURURE_MAX_V4;
+	}
+	for (j = 0; j < pd->maxkeybdnum; j++) {
+		for (i = 0; i < remotectl_button[j].nbuttons; i++) {
+			int scancode, usercode, pwrkey;
+
+			if (remotectl_button[j].key_table[i].keycode != KEY_POWER)
+				continue;
+			usercode = remotectl_button[j].usercode & 0xffff;
+			scancode = remotectl_button[j].key_table[i].scancode & 0xff;
+			DBG("usercode=%x, key=%x\n", usercode, scancode);
+			pwrkey  = usercode;
+			pwrkey |= (scancode << 24) | ((~scancode & 0xff) << 16);
+			DBG("pwrkey = %x\n", pwrkey);
+			writel_relaxed(pwrkey, reg + num * 4);
+			num++;
+			if (num >= num_max)
+				break;
+		}
+	}
+}
+
+static int rk_pwm_pwrkey_wakeup_init(struct platform_device *pdev)
+{
+	struct rkxx_remotectl_drvdata *ddata = platform_get_drvdata(pdev);
+	int ret = -1;
+
+	ddata->pwm_pwrkey_capture = 0;
+	if (ddata->pwm_data->pwm_version < 4) {
+		ret = rockchip_pwm_set_pwrmatch(pdev);
+		if (ret < 0)
+			goto end;
+	} else {
+		rockchip_pwm_set_pwrmatch_v4(ddata);
+	}
+	rockchip_pwrmatch_set_nec_param(ddata);
+	rockchip_pwrmatch_set_pwrkey(ddata);
+	ddata->pwm_pwrkey_capture = 1;
+end:
+	return ret;
+}
+
+static void rk_pwm_int_ctrl(struct platform_device *pdev, int work_mode)
+{
+	struct rkxx_remotectl_drvdata *ddata = platform_get_drvdata(pdev);
+	int val;
+	int pwm_id = ddata->remote_pwm_id;
+
+	if (pwm_id > 3)
+		return;
+	val = readl_relaxed(ddata->base + PWM_REG_INT_EN(pwm_id));
+	if (work_mode) {
+		val |= PWM_CH_INT_ENABLE(pwm_id);
+		DBG("pwm int enabled, value is 0x%x\n", val);
+		writel_relaxed(val, ddata->base + PWM_REG_INT_EN(pwm_id));
+	} else {
+		val &= ~PWM_CH_INT_ENABLE(pwm_id);
+		DBG("pwm int disabled, value is 0x%x\n", val);
+	}
+	writel_relaxed(val, ddata->base + PWM_REG_INT_EN(pwm_id));
+}
+
+static void rk_pwm_int_ctrl_v4(struct platform_device *pdev, int work_mode)
+{
+	struct rkxx_remotectl_drvdata *ddata = platform_get_drvdata(pdev);
+	int val;
+
+	if (work_mode) {
+		val = CAP_LPR_INT_EN(true) | CAP_HPR_INT_EN(true) | PWR_INT_EN(false);
+		writel_relaxed(val, ddata->base + PWM_REG_INT_EN_V4);
+	} else {
+		val = CAP_LPR_INT_EN(false) | CAP_HPR_INT_EN(false) | PWR_INT_EN(true);
+		writel_relaxed(val, ddata->base + PWM_REG_INT_EN_V4);
+	}
+}
+
+static int rk_pwm_remotectl_hw_init(struct platform_device *pdev)
+{
+	int val;
+	struct rkxx_remotectl_drvdata *ddata = platform_get_drvdata(pdev);
+
+	//1. disabled pwm
+	val = readl_relaxed(ddata->base + PWM_REG_CTRL);
+	val = (val & 0xFFFFFFFE) | PWM_DISABLE;
+	writel_relaxed(val, ddata->base + PWM_REG_CTRL);
+
+	//2. capture mode
+	val = readl_relaxed(ddata->base + PWM_REG_CTRL);
+	val = (val & 0xFFFFFFF9) | PWM_MODE_CAPTURE;
+	writel_relaxed(val, ddata->base + PWM_REG_CTRL);
+
+	//set clk div, clk div to 64
+	val = readl_relaxed(ddata->base + PWM_REG_CTRL);
+	val = (val & 0xFF0001FF) | PWM_DIV64;
+	writel_relaxed(val, ddata->base + PWM_REG_CTRL);
+
+	//4. enabled pwm int
+	ddata->pwm_data->funcs.int_ctrl(pdev, IR_WORK_MODE);
+
+	//5. enabled pwm
+	val = readl_relaxed(ddata->base + PWM_REG_CTRL);
+	val = (val & 0xFFFFFFFE) | PWM_ENABLE;
+	writel_relaxed(val, ddata->base + PWM_REG_CTRL);
+
+	return 0;
+}
+
+static int rk_pwm_remotectl_hw_init_v4(struct platform_device *pdev)
+{
+	int val;
+	struct rkxx_remotectl_drvdata *ddata = platform_get_drvdata(pdev);
+
+	//1. disabled pwm
+	val = PWM_EN(false) | PWM_CLK_EN(false);
+	writel_relaxed(val, ddata->base + PWM_REG_ENABLE_V4);
+
+	//2. capture mode
+	val = PWM_MODE(CAPTURE_MODE);
+	writel_relaxed(val, ddata->base + PWM_REG_CTRL_V4);
+
+	//3. set clk div, clk div to 64
+	val = CLK_SCALE(32);
+	writel_relaxed(val, ddata->base + PWM_REG_CLK_CTRL_V4);
+
+	//4. enabled pwm int
+	ddata->pwm_data->funcs.int_ctrl(pdev, IR_WORK_MODE);
+
+	//5. enabled pwm
+	val = PWM_EN(true) | PWM_CLK_EN(true);
+	writel_relaxed(val, ddata->base + PWM_REG_ENABLE_V4);
+	return 0;
+}
+
+static int rk_pwm_sip_wakeup_init(struct platform_device *pdev)
+{
+	struct rkxx_remotectl_drvdata *ddata = platform_get_drvdata(pdev);
+	struct device_node *np = pdev->dev.of_node;
+	struct irq_desc *desc;
+	int support_psci = 0;
+	int irq;
+	int hwirq;
+	int i, j;
+	int num;
+	int pwm_id;
+	int ret = -1;
+
+	if (of_property_read_u32(np, "remote_support_psci", &support_psci)) {
+		dev_info(&pdev->dev, "PWM Missing psci property in the DT.\n");
+		goto end;
+	}
+	DBG("support_psci=0x%x\n", support_psci);
+	if (!support_psci)
+		goto end;
+	irq = ddata->irq;
+	desc = irq_to_desc(irq);
+	if (!desc || !desc->irq_data.domain)
+		goto end;
+	hwirq = desc->irq_data.hwirq;
+	ret = sip_smc_remotectl_config(REMOTECTL_SET_IRQ, hwirq);
+	if (ret) {
+		dev_err(&pdev->dev, "set irq err, set support_psci to 0 !!\n");
+		/*
+		 * if atf doesn't support, return probe success to abandon atf
+		 * function and still use kernel pwm parse function
+		 */
+		goto end;
+	}
+	pwm_id = ddata->remote_pwm_id;
+	num = ddata->maxkeybdnum;
+	sip_smc_remotectl_config(REMOTECTL_SET_PWM_CH, pwm_id);
+	for (j = 0; j < num; j++) {
+		for (i = 0; i < remotectl_button[j].nbuttons; i++) {
+			int scancode, pwrkey;
+
+			if (remotectl_button[j].key_table[i].keycode
+			    != KEY_POWER)
+				continue;
+			scancode = remotectl_button[j].key_table[i].scancode;
+			DBG("usercode=%x, key=%x\n",
+			    remotectl_button[j].usercode, scancode);
+			pwrkey = (remotectl_button[j].usercode & 0xffff) << 16;
+			pwrkey |= (scancode & 0xff) << 8;
+			DBG("deliver: key=%x\n", pwrkey);
+			sip_smc_remotectl_config(REMOTECTL_SET_PWRKEY,
+							pwrkey);
+		}
+	}
+	sip_smc_remotectl_config(REMOTECTL_ENABLE, 1);
+	ddata->support_psci = support_psci;
+	DBG("rk pwm sip init end!\n");
+	return 0;
+end:
+	dev_info(&pdev->dev, "pwm sip wakeup config error!!\n");
+	ddata->support_psci = 0;
+	return ret;
+}
+
+static inline void rk_pwm_wakeup(struct input_dev *input)
+{
+	input_event(input, EV_KEY, KEY_POWER, 1);
+	input_event(input, EV_KEY, KEY_POWER, 0);
+	input_sync(input);
+}
+
+static const struct of_device_id rk_pwm_of_match[] = {
+	{ .compatible = "rockchip,remotectl-pwm", .data = &pwm_data_v1},
+	{ .compatible = "rockchip,remotectl-pwm-v4", .data = &pwm_data_v4},
+	{ }
+};
+
+MODULE_DEVICE_TABLE(of, rk_pwm_of_match);
+
+static int rk_pwm_probe(struct platform_device *pdev)
+{
+	struct rkxx_remotectl_drvdata *ddata;
+	const struct of_device_id *id;
+	struct device_node *np = pdev->dev.of_node;
+	struct resource *r;
+	struct input_dev *input;
+	struct clk *clk;
+	struct clk *p_clk;
+	struct cpumask cpumask;
+	int num;
+	int irq;
+	int ret;
+	int i, j;
+	int cpu_id;
+	int pwm_id;
+	int pwm_freq;
+	int count;
+
+	pr_err(".. rk pwm remotectl v2.0 init\n");
+	id = of_match_device(rk_pwm_of_match, &pdev->dev);
+	if (!id)
+		return -EINVAL;
+
+	r = platform_get_resource(pdev, IORESOURCE_MEM, 0);
+	if (!r) {
+		dev_err(&pdev->dev, "no memory resources defined\n");
+		return -ENODEV;
+	}
+
+	ddata = devm_kzalloc(&pdev->dev, sizeof(struct rkxx_remotectl_drvdata),
+			     GFP_KERNEL);
+	if (!ddata) {
+		return -ENOMEM;
+	}
+
+	ddata->pwm_data = id->data;
+	ddata->state = RMC_PRELOAD;
+	ddata->temp_period = 0;
+	ddata->base = devm_ioremap_resource(&pdev->dev, r);
+
+	if (IS_ERR(ddata->base))
+		return PTR_ERR(ddata->base);
+	count = of_property_count_strings(np, "clock-names");
+	if (count == 2) {
+		clk = devm_clk_get(&pdev->dev, "pwm");
+		p_clk = devm_clk_get(&pdev->dev, "pclk");
+	} else {
+		clk = devm_clk_get(&pdev->dev, NULL);
+		p_clk = clk;
+	}
+	if (IS_ERR(clk)) {
+		ret = PTR_ERR(clk);
+		if (ret != -EPROBE_DEFER)
+			dev_err(&pdev->dev, "Can't get bus clk: %d\n", ret);
+		return ret;
+	}
+	if (IS_ERR(p_clk)) {
+		ret = PTR_ERR(p_clk);
+		if (ret != -EPROBE_DEFER)
+			dev_err(&pdev->dev, "Can't get periph clk: %d\n", ret);
+		return ret;
+	}
+	ret = clk_prepare_enable(clk);
+	if (ret) {
+		dev_err(&pdev->dev, "Can't enable bus clk: %d\n", ret);
+		return ret;
+	}
+	ret = clk_prepare_enable(p_clk);
+	if (ret) {
+		dev_err(&pdev->dev, "Can't enable bus periph clk: %d\n", ret);
+		goto error_clk;
+	}
+	platform_set_drvdata(pdev, ddata);
+	ddata->dev = &pdev->dev;
+	num = rk_remotectl_get_irkeybd_count(pdev);
+	if (num == 0) {
+		pr_err("remotectl: no ir keyboard add in dts!!\n");
+		ret = -EINVAL;
+		goto error_pclk;
+	}
+	ddata->maxkeybdnum = num;
+	remotectl_button = devm_kzalloc(&pdev->dev,
+					num * sizeof(struct rkxx_remotectl_button),
+					GFP_KERNEL);
+	if (!remotectl_button) {
+		pr_err("failed to malloc remote button memory\n");
+		ret = -ENOMEM;
+		goto error_pclk;
+	}
+	input = devm_input_allocate_device(&pdev->dev);
+	if (!input) {
+		pr_err("failed to allocate input device\n");
+		ret = -ENOMEM;
+		goto error_pclk;
+	}
+	input->name = pdev->name;
+	input->phys = "gpio-keys/remotectl";
+	input->dev.parent = &pdev->dev;
+	input->id.bustype = BUS_HOST;
+	input->id.vendor = 0x524b;
+	input->id.product = 0x0006;
+	input->id.version = 0x0100;
+	ddata->input = input;
+	irq = platform_get_irq(pdev, 0);
+	if (irq < 0) {
+		dev_err(&pdev->dev, "cannot find IRQ\n");
+		goto error_pclk;
+	}
+	ddata->irq = irq;
+	ddata->wakeup = 1;
+
+	if (ddata->pwm_data->pwm_version < 4) {
+		of_property_read_u32(np, "remote_pwm_id", &pwm_id);
+		pwm_id %= 4;
+		ddata->remote_pwm_id = pwm_id;
+		if (pwm_id > 3) {
+			dev_err(&pdev->dev, "pwm id error\n");
+			goto error_pclk;
+		}
+		DBG("remotectl: remote pwm id=0x%x\n", pwm_id);
+	}
+	of_property_read_u32(np, "handle_cpu_id", &cpu_id);
+	ddata->handle_cpu_id = cpu_id;
+	DBG("remotectl: handle cpu id=0x%x\n", cpu_id);
+	rk_remotectl_parse_ir_keys(pdev);
+	tasklet_init(&ddata->remote_tasklet, rk_pwm_remotectl_do_something,
+		     (unsigned long)ddata);
+	for (j = 0; j < num; j++) {
+		DBG("remotectl probe j = 0x%x\n", j);
+		for (i = 0; i < remotectl_button[j].nbuttons; i++) {
+			int keycode;
+
+			keycode = remotectl_button[j].key_table[i].keycode;
+			input_set_capability(input, EV_KEY, keycode);
+		}
+	}
+	ret = input_register_device(input);
+	if (ret)
+		dev_err(&pdev->dev, "register input device err, ret=%d\n", ret);
+	input_set_capability(input, EV_KEY, KEY_WAKEUP);
+	device_init_wakeup(&pdev->dev, 1);
+	enable_irq_wake(irq);
+	timer_setup(&ddata->timer, rk_pwm_remotectl_timer, 0);
+	wake_lock_init(&ddata->remotectl_wake_lock,
+		       WAKE_LOCK_SUSPEND, "rockchip_pwm_remote");
+	cpumask_clear(&cpumask);
+	cpumask_set_cpu(cpu_id, &cpumask);
+	irq_set_affinity_hint(irq, &cpumask);
+	ret = devm_request_irq(&pdev->dev, irq, ddata->pwm_data->funcs.irq_handler,
+			       IRQF_NO_SUSPEND, "rk_pwm_irq", ddata);
+	if (ret) {
+		dev_err(&pdev->dev, "cannot claim IRQ %d\n", irq);
+		goto error_irq;
+	}
+	pwm_freq = clk_get_rate(clk) / 64;
+	ddata->pwm_freq_nstime = 1000000000 / pwm_freq;
+	ddata->pwm_data->funcs.init_hw(pdev);
+	ret = rk_pwm_pwrkey_wakeup_init(pdev);
+	if (!ret) {
+		dev_info(&pdev->dev, "Controller support pwrkey capture\n");
+		goto end;
+	}
+	ret = rk_pwm_sip_wakeup_init(pdev);
+	if (ret)
+		dev_info(&pdev->dev, "Donot support ATF Wakeup\n");
+	else
+		dev_info(&pdev->dev, "Support ATF Wakeup\n");
+	DBG("rk pwm remotectl init end!\n");
+end:
+	return 0;
+error_irq:
+error_pclk:
+	clk_unprepare(p_clk);
+error_clk:
+	clk_unprepare(clk);
+	return ret;
+}
+
+static int rk_pwm_remove(struct platform_device *pdev)
+{
+	return 0;
+}
+
+static int remotectl_suspend(struct device *dev)
+{
+	int cpu = 0;
+	struct cpumask cpumask;
+	struct platform_device *pdev = to_platform_device(dev);
+	struct rkxx_remotectl_drvdata *ddata = platform_get_drvdata(pdev);
+
+	if (ddata->pwm_pwrkey_capture) {
+		ddata->pwrkey_wakeup = 0;
+		ddata->pwm_data->funcs.int_ctrl(pdev, IR_SUSPEND_MODE);
+	}
+	cpumask_clear(&cpumask);
+	cpumask_set_cpu(cpu, &cpumask);
+	irq_set_affinity_hint(ddata->irq, &cpumask);
+	return 0;
+}
+
+#ifdef CONFIG_PM
+static int remotectl_resume(struct device *dev)
+{
+	struct cpumask cpumask;
+	struct platform_device *pdev = to_platform_device(dev);
+	struct rkxx_remotectl_drvdata *ddata = platform_get_drvdata(pdev);
+	int state;
+
+	cpumask_clear(&cpumask);
+	cpumask_set_cpu(ddata->handle_cpu_id, &cpumask);
+	irq_set_affinity_hint(ddata->irq, &cpumask);
+	if (ddata->support_psci) {
+		/*
+		 * loop wakeup state
+		 */
+		state = sip_smc_remotectl_config(
+					REMOTECTL_GET_WAKEUP_STATE, 0);
+		if (state == REMOTECTL_PWRKEY_WAKEUP)
+			rk_pwm_wakeup(ddata->input);
+	}  else if (ddata->pwm_pwrkey_capture) {
+		ddata->pwm_data->funcs.int_ctrl(pdev, IR_WORK_MODE);
+		if (ddata->pwrkey_wakeup == 0)
+			return 0;
+		ddata->pwrkey_wakeup = 0;
+		rk_pwm_wakeup(ddata->input);
+	}
+
+	return 0;
+}
+
+static const struct dev_pm_ops remotectl_pm_ops = {
+	.suspend_late = remotectl_suspend,
+	.resume_early = remotectl_resume,
+};
+#endif
+
+static void rk_pwm_remotectl_shutdown(struct platform_device *pdev)
+{
+	remotectl_suspend(&pdev->dev);
+}
+
+static struct platform_driver rk_pwm_driver = {
+	.driver = {
+		.name = "remotectl-pwm",
+		.of_match_table = rk_pwm_of_match,
+#ifdef CONFIG_PM
+		.pm = &remotectl_pm_ops,
+#endif
+	},
+	.remove = rk_pwm_remove,
+	.shutdown = rk_pwm_remotectl_shutdown,
+};
+
+module_platform_driver_probe(rk_pwm_driver, rk_pwm_probe);
+
+MODULE_LICENSE("GPL");
diff -ruN a/drivers/input/remotectl/rockchip_pwm_remotectl.h b/drivers/input/remotectl/rockchip_pwm_remotectl.h
--- a/drivers/input/remotectl/rockchip_pwm_remotectl.h	1969-12-31 16:00:00.000000000 -0800
+++ b/drivers/input/remotectl/rockchip_pwm_remotectl.h	2026-06-23 04:06:29.908977447 -0700
@@ -0,0 +1,222 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+
+#ifndef __RKXX_PWM_REMOTECTL_H__
+#define __RKXX_PWM_REMOTECTL_H__
+#include <linux/input.h>
+
+#define MAX_NUM_KEYS			60
+#define PWM_PWR_KEY_CAPURURE_MAX	10
+#define PWM_PWR_KEY_CAPURURE_MAX_V4	16
+
+/* PWM0 registers  */
+#define PWM_REG_CNTR			0x00  /* Counter Register */
+#define PWM_REG_HPR			0x04  /* Period Register */
+#define PWM_REG_LPR			0x08  /* Duty Cycle Register */
+#define PWM_REG_CTRL			0x0c  /* Control Register */
+#define PWM3_REG_INTSTS			0x10  /* Interrupt Status Refister For Pwm3*/
+#define PWM2_REG_INTSTS			0x20  /* Interrupt Status Refister For Pwm2*/
+#define PWM1_REG_INTSTS			0x30  /* Interrupt Status Refister For Pwm1*/
+#define PWM0_REG_INTSTS			0x40  /* Interrupt Status Refister For Pwm0*/
+#define PWM3_REG_INT_EN			0x14  /* Interrupt Enable Refister For Pwm3*/
+#define PWM2_REG_INT_EN			0x24  /* Interrupt Enable Refister For Pwm2*/
+#define PWM1_REG_INT_EN			0x34  /* Interrupt Enable Refister For Pwm1*/
+#define PWM0_REG_INT_EN			0x44  /* Interrupt Enable Refister For Pwm0*/
+
+/*REG_CTRL bits definitions*/
+#define PWM_ENABLE			(1 << 0)
+#define PWM_DISABLE			(0 << 0)
+
+/*operation mode*/
+#define PWM_MODE_ONESHOT		(0x00 << 1)
+#define PWM_MODE_CONTINUMOUS		(0x01 << 1)
+#define PWM_MODE_CAPTURE		(0x02 << 1)
+
+/*duty cycle output polarity*/
+#define PWM_DUTY_POSTIVE		(0x01 << 3)
+#define PWM_DUTY_NEGATIVE		(0x00 << 3)
+
+/*incative state output polarity*/
+#define PWM_INACTIVE_POSTIVE		(0x01 << 4)
+#define PWM_INACTIVE_NEGATIVE		(0x00 << 4)
+
+/*clock source select*/
+#define PWM_CLK_SCALE			(1 << 9)
+#define PWM_CLK_NON_SCALE		(0 << 9)
+
+#define PWM_CH0_INT			(1 << 0)
+#define PWM_CH1_INT			(1 << 1)
+#define PWM_CH2_INT			(1 << 2)
+#define PWM_CH3_INT			(1 << 3)
+#define PWM_PWR_KEY_INT			(1 << 7)
+
+#define PWM_CH0_POL			(1 << 8)
+#define PWM_CH1_POL			(1 << 9)
+#define PWM_CH2_POL			(1 << 10)
+#define PWM_CH3_POL			(1 << 11)
+
+#define PWM_CH0_INT_ENABLE		(1 << 0)
+#define PWM_CH0_INT_DISABLE		(0 << 0)
+
+#define PWM_CH1_INT_ENABLE		(1 << 1)
+#define PWM_CH1_INT_DISABLE		(0 << 1)
+
+#define PWM_CH2_INT_ENABLE		(1 << 2)
+#define PWM_CH2_INT_DISABLE		(0 << 2)
+
+#define PWM_CH3_INT_ENABLE		(1 << 3)
+#define PWM_CH3_INT_DISABLE		(0 << 3)
+
+#define PWM_INT_ENABLE			1
+#define PWM_INT_DISABLE			0
+
+/*prescale factor*/
+#define PWMCR_MIN_PRESCALE			0x00
+#define PWMCR_MAX_PRESCALE			0x07
+
+#define PWMDCR_MIN_DUTY				0x0001
+#define PWMDCR_MAX_DUTY				0xFFFF
+
+#define PWMPCR_MIN_PERIOD			0x0001
+#define PWMPCR_MAX_PERIOD			0xFFFF
+
+#define PWMPCR_MIN_PERIOD			0x0001
+#define PWMPCR_MAX_PERIOD			0xFFFF
+
+
+enum pwm_div {
+	PWM_DIV1	= (0x0 << 12),
+	PWM_DIV2	= (0x1 << 12),
+	PWM_DIV4	= (0x2 << 12),
+	PWM_DIV8	= (0x3 << 12),
+	PWM_DIV16	= (0x4 << 12),
+	PWM_DIV32	= (0x5 << 12),
+	PWM_DIV64	= (0x6 << 12),
+	PWM_DIV128	= (0x7 << 12),
+};
+
+/*
+ * regs for pwm v4
+ */
+#define	PWM_REG_VERSION_V4			0x0
+#define	PWM_REG_ENABLE_V4			0x4
+#define	PWM_REG_CLK_CTRL_V4			0x8
+#define	PWM_REG_CTRL_V4				0xC
+#define	PWM_REG_HPR_V4				0x2C
+#define	PWM_REG_LPR_V4				0x30
+#define	PWM_REG_INTSTS_V4			0x70
+#define	PWM_REG_INT_EN_V4			0x74
+#define	PWM_REG_MATCH_ARBITER_V4		0x100
+#define	PWM_REG_MATCH_CTRL_V4			0x104
+#define	PWM_REG_MATCH_LPRE_V4			0x108
+#define	PWM_REG_MATCH_HPRE_V4			0x10C
+#define	PWM_REG_MATCH_LD_V4			0x110
+#define	PWM_REG_MATCH_HD_ZERO_V4		0x114
+#define	PWM_REG_MATCH_HD_ONE_V4			0x118
+#define	PWM_REG_CAPTURE_VALUE0_V4		0x11C
+#define	PWM_REG_CAPTURE_VALUE_V4		0x15C
+
+
+#define HIWORD_UPDATE(v, l, h)	(((v) << (l)) | (GENMASK(h, l) << 16))
+
+/* VERSION_ID */
+#define	CHANNLE_INDEX_SHIFT		4
+#define	CHANNLE_INDEX_MASK		(0xf << CHANNLE_INDEX_SHIFT)
+
+/* PWM_CTRL */
+#define	PWM_MODE(v)			HIWORD_UPDATE(v, 0, 1)
+#define	CAPTURE_MODE			2
+#define	CLK_SCALE(v)			HIWORD_UPDATE(v, 4, 12)
+
+/* INTSTS */
+#define	CAP_LPR_INTSTS_SHIFT		0
+#define	CAP_HPR_INTSTS_SHIFT		1
+#define	PWR_INTSTS_SHIFT		5
+#define	CAP_LPR_INT			BIT(CAP_LPR_INTSTS_SHIFT)
+#define	CAP_HPR_INT			BIT(CAP_HPR_INTSTS_SHIFT)
+#define	PWR_INT				BIT(PWR_INTSTS_SHIFT)
+
+/* INT_EN */
+#define	CAP_LPR_INT_EN(v)		HIWORD_UPDATE(v, 0, 0)
+#define	CAP_HPR_INT_EN(v)		HIWORD_UPDATE(v, 1, 1)
+#define	PWR_INT_EN(v)			HIWORD_UPDATE(v, 5, 5)
+
+/* INT_MASK */
+#define	CAP_LPR_INT_MASK(v)		HIWORD_UPDATE(v, 0, 0)
+#define	CAP_HPR_INT_MASK(v)		HIWORD_UPDATE(v, 1, 1)
+#define	PWR_INT_MASK(v)			HIWORD_UPDATE(v, 5, 5)
+
+/* PWRMATCH_ARBITER */
+#define	PWRMATCH_GRANT_SHIFT		0
+#define	PWRMATCH_READ_LOCK_SHIFT	16
+#define	PWRKEY_EN(v)			HIWORD_UPDATE(v, 0, 0)
+#define	PWM_CLK_EN(v)			HIWORD_UPDATE(v, 0, 0)
+#define	PWM_EN(v)			HIWORD_UPDATE(v, 1, 1)
+
+
+/*  NEC Protocol */
+#define	RK_PWM_TIME_PRE_MIN		4000
+#define	RK_PWM_TIME_PRE_MAX		5000
+
+#define	RK_PWM_TIME_PRE_MIN_LOW		8000
+#define	RK_PWM_TIME_PRE_MAX_LOW		10000
+
+#define	RK_PWM_TIME_BIT0_MIN		390
+#define	RK_PWM_TIME_BIT0_MAX		730
+
+#define	RK_PWM_TIME_BIT1_MIN		1300
+#define	RK_PWM_TIME_BIT1_MAX		2000
+
+#define	RK_PWM_TIME_BIT_MIN_LOW		390
+#define	RK_PWM_TIME_BIT_MAX_LOW		730
+
+#define	RK_PWM_TIME_RPT_MIN		2000
+#define	RK_PWM_TIME_RPT_MAX		2500
+
+#define	RK_PWM_TIME_SEQ1_MIN		95000
+#define	RK_PWM_TIME_SEQ1_MAX		98000
+
+#define	RK_PWM_TIME_SEQ2_MIN		30000
+#define	RK_PWM_TIME_SEQ2_MAX		55000
+
+
+#define	PWM_REG_INTSTS(n)		((3 - (n)) * 0x10 + 0x10)
+#define	PWM_REG_INT_EN(n)		((3 - (n)) * 0x10 + 0x14)
+#define	RK_PWM_VERSION_ID(n)		((3 - (n)) * 0x10 + 0x2c)
+#define	PWM_REG_PWRMATCH_CTRL(n)	((3 - (n)) * 0x10 + 0x50)
+#define	PWM_REG_PWRMATCH_LPRE(n)	((3 - (n)) * 0x10 + 0x54)
+#define	PWM_REG_PWRMATCH_HPRE(n)	((3 - (n)) * 0x10 + 0x58)
+#define	PWM_REG_PWRMATCH_LD(n)		((3 - (n)) * 0x10 + 0x5C)
+#define	PWM_REG_PWRMATCH_HD_ZERO(n)	((3 - (n)) * 0x10 + 0x60)
+#define	PWM_REG_PWRMATCH_HD_ONE(n)	((3 - (n)) * 0x10 + 0x64)
+#define	PWM_PWRMATCH_VALUE(n)		((3 - (n)) * 0x10 + 0x68)
+#define	PWM_PWRCAPTURE_VALUE(n)		((3 - (n)) * 0x10 + 0x9c)
+
+#define	PWM_CH_INT(n)			BIT(n)
+#define	PWM_CH_POL(n)			BIT(n+8)
+
+#define	PWM_CH_INT_ENABLE(n)		BIT(n)
+#define	PWM_PWR_INT_ENABLE		BIT(7)
+#define	CH3_PWRKEY_ENABLE		BIT(3)
+
+
+typedef enum _RMC_STATE {
+	RMC_IDLE,
+	RMC_PRELOAD,
+	RMC_USERCODE,
+	RMC_GETDATA,
+	RMC_SEQUENCE,
+} eRMC_STATE;
+
+enum {
+	IR_SUSPEND_MODE = 0,
+	IR_WORK_MODE
+};
+
+struct RKxx_remotectl_platform_data {
+	int nbuttons;
+	int rep;
+	int timer;
+	int wakeup;
+};
+
+#endif
diff -ruN a/include/linux/rockchip/rockchip_sip.h b/include/linux/rockchip/rockchip_sip.h
--- a/include/linux/rockchip/rockchip_sip.h	1969-12-31 16:00:00.000000000 -0800
+++ b/include/linux/rockchip/rockchip_sip.h	2026-06-23 04:09:44.535237027 -0700
@@ -0,0 +1,564 @@
+/* Copyright (c) 2016, Fuzhou Rockchip Electronics Co., Ltd
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2 and
+ * only version 2 as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+#ifndef __ROCKCHIP_SIP_H
+#define __ROCKCHIP_SIP_H
+
+#include <linux/arm-smccc.h>
+#include <linux/io.h>
+
+/* SMC function IDs for SiP Service queries, compatible with kernel-3.10 */
+#define SIP_ATF_VERSION			0x82000001
+#define SIP_ACCESS_REG			0x82000002
+#define SIP_SUSPEND_MODE		0x82000003
+#define SIP_PENDING_CPUS		0x82000004
+#define SIP_UARTDBG_CFG			0x82000005
+#define SIP_UARTDBG_CFG64		0xc2000005
+#define SIP_MCU_EL3FIQ_CFG		0x82000006
+#define SIP_ACCESS_CHIP_STATE64		0xc2000006
+#define SIP_SECURE_MEM_CONFIG		0x82000007
+#define SIP_ACCESS_CHIP_EXTRA_STATE64	0xc2000007
+#define SIP_DRAM_CONFIG			0x82000008
+#define SIP_SHARE_MEM			0x82000009
+#define SIP_SIP_VERSION			0x8200000a
+#define SIP_REMOTECTL_CFG		0x8200000b
+#define PSCI_SIP_VPU_RESET		0x8200000c
+#define SIP_BUS_CFG			0x8200000d
+#define SIP_LAST_LOG			0x8200000e
+#define SIP_ACCESS_MEM_OS_REG		0x8200000f
+#define SIP_SCMI_AGENT0			0x82000010
+#define SIP_SCMI_AGENT1			0x82000011
+#define SIP_SCMI_AGENT2			0x82000012
+#define SIP_SCMI_AGENT3			0x82000013
+#define SIP_SCMI_AGENT4			0x82000014
+#define SIP_SCMI_AGENT5			0x82000015
+#define SIP_SCMI_AGENT6			0x82000016
+#define SIP_SCMI_AGENT7			0x82000017
+#define SIP_SCMI_AGENT8			0x82000018
+#define SIP_SCMI_AGENT9			0x82000019
+#define SIP_SCMI_AGENT10		0x8200001a
+#define SIP_SCMI_AGENT11		0x8200001b
+#define SIP_SCMI_AGENT12		0x8200001c
+#define SIP_SCMI_AGENT13		0x8200001d
+#define SIP_SCMI_AGENT14		0x8200001e
+#define SIP_SCMI_AGENT15		0x8200001f
+#define SIP_SDEI_FIQ_DBG_SWITCH_CPU	0x82000020
+#define SIP_SDEI_FIQ_DBG_GET_EVENT_ID	0x82000021
+#define RK_SIP_AMP_CFG			0x82000022
+#define RK_SIP_FIQ_CTRL			0x82000024
+#define SIP_HDCP_CONFIG			0x82000025
+#define SIP_WDT_CFG			0x82000026
+#define SIP_HDMIRX_CFG			0x82000027
+#define SIP_MCU_CFG			0x82000028
+#define SIP_PVTPLL_CFG			0x82000029
+
+#define TRUSTED_OS_HDCPKEY_INIT		0xB7000003
+
+/* Rockchip Sip version */
+#define SIP_IMPLEMENT_V1                (1)
+#define SIP_IMPLEMENT_V2                (2)
+
+/* Trust firmware version */
+#define ATF_VER_MAJOR(ver)		(((ver) >> 16) & 0xffff)
+#define ATF_VER_MINOR(ver)		(((ver) >> 0) & 0xffff)
+
+/* SIP_ACCESS_REG: read or write */
+#define SECURE_REG_RD			0x0
+#define SECURE_REG_WR			0x1
+
+/* Fiq debugger share memory: 8KB enough */
+#define FIQ_UARTDBG_PAGE_NUMS		2
+#define FIQ_UARTDBG_SHARE_MEM_SIZE	((FIQ_UARTDBG_PAGE_NUMS) * 4096)
+
+/* Error return code */
+#define IS_SIP_ERROR(x)			(!!(x))
+
+#define SIP_RET_SUCCESS			0
+#define SIP_RET_SMC_UNKNOWN		-1
+#define SIP_RET_NOT_SUPPORTED		-2
+#define SIP_RET_INVALID_PARAMS		-3
+#define SIP_RET_INVALID_ADDRESS		-4
+#define SIP_RET_DENIED			-5
+#define SIP_RET_SET_RATE_TIMEOUT	-6
+
+/* SIP_UARTDBG_CFG64 call types */
+#define UARTDBG_CFG_INIT		0xf0
+#define UARTDBG_CFG_OSHDL_TO_OS		0xf1
+#define UARTDBG_CFG_OSHDL_CPUSW		0xf3
+#define UARTDBG_CFG_OSHDL_DEBUG_ENABLE	0xf4
+#define UARTDBG_CFG_OSHDL_DEBUG_DISABLE	0xf5
+#define UARTDBG_CFG_PRINT_PORT		0xf7
+#define UARTDBG_CFG_FIQ_ENABEL		0xf8
+#define UARTDBG_CFG_FIQ_DISABEL		0xf9
+
+/* SIP_SUSPEND_MODE32 call types */
+#define SUSPEND_MODE_CONFIG		0x01
+#define WKUP_SOURCE_CONFIG		0x02
+#define PWM_REGULATOR_CONFIG		0x03
+#define GPIO_POWER_CONFIG		0x04
+#define SUSPEND_DEBUG_ENABLE		0x05
+#define APIOS_SUSPEND_CONFIG		0x06
+#define VIRTUAL_POWEROFF		0x07
+#define SUSPEND_WFI_TIME_MS		0x08
+#define LINUX_PM_STATE			0x09
+#define SUSPEND_IO_RET_CONFIG		0x0a
+#define SLEEP_PIN_CONFIG		0x0b
+
+/* SIP_REMOTECTL_CFG call types */
+#define	REMOTECTL_SET_IRQ		0xf0
+#define REMOTECTL_SET_PWM_CH		0xf1
+#define REMOTECTL_SET_PWRKEY		0xf2
+#define REMOTECTL_GET_WAKEUP_STATE	0xf3
+#define REMOTECTL_ENABLE		0xf4
+/* wakeup state */
+#define REMOTECTL_PWRKEY_WAKEUP		0xdeadbeaf
+
+/* SIP_ACCESS_MEM_OS_REG child configs */
+enum {
+	RK_MEM_OS_REG_READ = 0,
+	RK_MEM_OS_REG_WRITE,
+};
+
+/* SIP_MCU_CFG child configs, MCU ID */
+enum {
+	RK_BUS_MCU,
+	RK_PMU_MCU,
+	RK_DDR_MCU,
+	RK_NPU_MCU,
+};
+
+#define RK_SIP_MCU_ID(type, id)		((type) << 8 | id)
+
+#define RK_SIP_CFG_BUSMCU_0_ID		RK_SIP_MCU_ID(RK_BUS_MCU, 0)
+#define RK_SIP_CFG_BUSMCU_1_ID		RK_SIP_MCU_ID(RK_BUS_MCU, 1)
+#define RK_SIP_CFG_PMUMCU_0_ID		RK_SIP_MCU_ID(RK_PMU_MCU, 0)
+#define RK_SIP_CFG_DDRMCU_0_ID		RK_SIP_MCU_ID(RK_DDR_MCU, 0)
+#define RK_SIP_CFG_NPUMCU_0_ID		RK_SIP_MCU_ID(RK_NPU_MCU, 0)
+
+/* SIP_MCU_CFG child configs */
+#define CONFIG_MCU_CODE_START_ADDR	0x01
+#define CONFIG_MCU_EXPERI_START_ADDR	0x02
+#define CONFIG_MCU_SRAM_START_ADDR	0x03
+#define CONFIG_MCU_EXSRAM_START_ADDR	0x04
+
+struct dram_addrmap_info {
+	u64 ch_mask[2];
+	u64 bk_mask[4];
+	u64 bg_mask[2];
+	u64 cs_mask[2];
+	u32 reserved[20];
+	u32 bank_bit_first;
+	u32 bank_bit_mask;
+};
+
+/* AMP Ctrl */
+enum {
+	RK_AMP_SUB_FUNC_CFG_MODE = 0,
+	RK_AMP_SUB_FUNC_BOOT_ARG01,
+	RK_AMP_SUB_FUNC_BOOT_ARG23,
+	RK_AMP_SUB_FUNC_REQ_CPU_OFF,
+	RK_AMP_SUB_FUNC_GET_CPU_STATUS,
+	RK_AMP_SUB_FUNC_RSV, /* for RTOS */
+	RK_AMP_SUB_FUNC_CPU_ON,
+	RK_AMP_SUB_FUNC_END,
+};
+
+enum {
+	FIRMWARE_NONE,
+	FIRMWARE_TEE_32BIT,
+	FIRMWARE_ATF_32BIT,
+	FIRMWARE_ATF_64BIT,
+	FIRMWARE_END,
+};
+
+/* Share mem page types */
+typedef enum {
+	SHARE_PAGE_TYPE_INVALID = 0,
+	SHARE_PAGE_TYPE_UARTDBG,
+	SHARE_PAGE_TYPE_DDR,
+	SHARE_PAGE_TYPE_DDRDBG,
+	SHARE_PAGE_TYPE_DDRECC,
+	SHARE_PAGE_TYPE_DDRFSP,
+	SHARE_PAGE_TYPE_DDR_ADDRMAP,
+	SHARE_PAGE_TYPE_LAST_LOG,
+	SHARE_PAGE_TYPE_HDCP,
+	SHARE_PAGE_TYPE_SLEEP,
+	SHARE_PAGE_TYPE_MAX,
+} share_page_type_t;
+
+/* fiq control sub func */
+enum {
+	RK_SIP_FIQ_CTRL_FIQ_EN = 1,
+	RK_SIP_FIQ_CTRL_FIQ_DIS,
+	RK_SIP_FIQ_CTRL_SET_AFF
+};
+
+/* hdcp function types */
+enum {
+	HDCP_FUNC_STORAGE_INCRYPT = 1,
+	HDCP_FUNC_KEY_LOAD,
+	HDCP_FUNC_ENCRYPT_MODE
+};
+
+/* support hdcp device list */
+enum {
+	DP_TX0,
+	DP_TX1,
+	EDP_TX0,
+	EDP_TX1,
+	HDMI_TX0,
+	HDMI_TX1,
+	HDMI_RX,
+	MAX_DEVICE,
+};
+
+/* SIP_WDT_CONFIG call types  */
+enum {
+	WDT_START = 0,
+	WDT_STOP = 1,
+	WDT_PING = 2,
+};
+
+/* SIP_HDMIRX_CONFIG child configs */
+enum {
+	HDMIRX_AUTO_TOUCH_EN = 0,
+	HDMIRX_REG_PRE_FETCH = 1,
+	HDMIRX_INFO_NOTIFY = 2,
+};
+
+/* SIP_PVTPLL_CFG child configs */
+enum {
+	PVTPLL_GET_INFO = 0,
+	PVTPLL_ADJUST_TABLE = 1,
+	PVTPLL_LOW_TEMP = 2,
+};
+
+struct pt_regs;
+typedef void (*sip_fiq_debugger_uart_irq_tf_cb_t)(struct pt_regs *_pt_regs, unsigned long cpu);
+
+/*
+ * Rules: struct arm_smccc_res contains result and data, details:
+ *
+ * a0: error code(0: success, !0: error);
+ * a1~a3: data
+ */
+#if IS_REACHABLE(CONFIG_ROCKCHIP_SIP)
+struct arm_smccc_res sip_smc_get_atf_version(void);
+struct arm_smccc_res sip_smc_get_sip_version(void);
+struct arm_smccc_res sip_smc_dram(u32 arg0, u32 arg1, u32 arg2);
+struct arm_smccc_res sip_smc_request_share_mem(u32 page_num,
+					       share_page_type_t page_type);
+struct arm_smccc_res sip_smc_mcu_el3fiq(u32 arg0, u32 arg1, u32 arg2);
+struct arm_smccc_res sip_smc_vpu_reset(u32 arg0, u32 arg1, u32 arg2);
+struct arm_smccc_res sip_smc_get_suspend_info(u32 info);
+struct arm_smccc_res sip_smc_lastlog_request(void);
+
+int sip_smc_set_suspend_mode(u32 ctrl, u32 config1, u32 config2);
+int sip_smc_virtual_poweroff(void);
+int sip_smc_remotectl_config(u32 func, u32 data);
+
+int sip_smc_secure_reg_write(u32 addr_phy, u32 val);
+u32 sip_smc_secure_reg_read(u32 addr_phy);
+struct arm_smccc_res sip_smc_bus_config(u32 arg0, u32 arg1, u32 arg2);
+struct dram_addrmap_info *sip_smc_get_dram_map(void);
+int sip_smc_access_mem_os_reg(u32 func, u32 id, u32 *val);
+int sip_smc_amp_config(u32 sub_func_id, u32 arg1, u32 arg2, u32 arg3);
+struct arm_smccc_res sip_smc_get_amp_info(u32 sub_func_id, u32 arg1);
+struct arm_smccc_res sip_smc_get_pvtpll_info(u32 sub_func_id, u32 arg1);
+struct arm_smccc_res sip_smc_pvtpll_config(u32 sub_func_id, u32 arg1, u32 arg2,
+					   u32 arg3, u32 arg4, u32 arg5, u32 arg6);
+
+void __iomem *sip_hdcp_request_share_memory(int id);
+struct arm_smccc_res sip_hdcp_config(u32 arg0, u32 arg1, u32 arg2);
+ulong sip_cpu_logical_map_mpidr(u32 cpu);
+/***************************fiq debugger **************************************/
+void sip_fiq_debugger_enable_fiq(bool enable, uint32_t tgt_cpu);
+void sip_fiq_debugger_enable_debug(bool enable);
+int sip_fiq_debugger_uart_irq_tf_init(u32 irq_id, sip_fiq_debugger_uart_irq_tf_cb_t callback_fn);
+int sip_fiq_debugger_set_print_port(u32 port_phyaddr, u32 baudrate);
+int sip_fiq_debugger_request_share_memory(void);
+int sip_fiq_debugger_get_target_cpu(void);
+int sip_fiq_debugger_switch_cpu(u32 cpu);
+int sip_fiq_debugger_sdei_switch_cpu(u32 cur_cpu, u32 target_cpu, u32 flag);
+int sip_fiq_debugger_is_enabled(void);
+int sip_fiq_debugger_sdei_get_event_id(u32 *fiq, u32 *sw_cpu, u32 *flag);
+int sip_fiq_control(u32 sub_func, u32 irq, unsigned long data);
+int sip_wdt_config(u32 sub_func, u32 arg1, u32 arg2, u32 arg3);
+int sip_hdmirx_config(u32 sub_func, u32 arg1, u32 arg2, u32 arg3);
+int sip_hdcpkey_init(u32 hdcp_id);
+int sip_smc_mcu_config(unsigned long mcu_id, unsigned long func, unsigned long arg2);
+#else
+static inline struct arm_smccc_res sip_smc_get_atf_version(void)
+{
+	struct arm_smccc_res tmp = { .a0 = SIP_RET_NOT_SUPPORTED };
+	return tmp;
+}
+
+static inline struct arm_smccc_res sip_smc_get_sip_version(void)
+{
+	struct arm_smccc_res tmp = { .a0 = SIP_RET_NOT_SUPPORTED };
+	return tmp;
+}
+
+static inline struct arm_smccc_res sip_smc_dram(u32 arg0, u32 arg1, u32 arg2)
+{
+	struct arm_smccc_res tmp = { .a0 = SIP_RET_NOT_SUPPORTED };
+	return tmp;
+}
+
+static inline struct arm_smccc_res sip_smc_request_share_mem
+			(u32 page_num, share_page_type_t page_type)
+{
+	struct arm_smccc_res tmp = { .a0 = SIP_RET_NOT_SUPPORTED };
+	return tmp;
+}
+
+static inline struct arm_smccc_res sip_smc_mcu_el3fiq
+			(u32 arg0, u32 arg1, u32 arg2)
+{
+	struct arm_smccc_res tmp = { .a0 = SIP_RET_NOT_SUPPORTED };
+	return tmp;
+}
+
+static inline struct arm_smccc_res
+sip_smc_vpu_reset(u32 arg0, u32 arg1, u32 arg2)
+{
+	struct arm_smccc_res tmp = { .a0 = SIP_RET_NOT_SUPPORTED };
+	return tmp;
+}
+
+static inline struct arm_smccc_res sip_smc_get_suspend_info(u32 info)
+{
+	struct arm_smccc_res tmp = { .a0 = SIP_RET_NOT_SUPPORTED };
+	return tmp;
+}
+
+static inline struct arm_smccc_res sip_smc_lastlog_request(void)
+{
+	struct arm_smccc_res tmp = { .a0 = SIP_RET_NOT_SUPPORTED };
+	return tmp;
+}
+
+static inline int sip_smc_set_suspend_mode(u32 ctrl, u32 config1, u32 config2)
+{
+	return 0;
+}
+
+static inline int sip_smc_virtual_poweroff(void) { return 0; }
+static inline int sip_smc_remotectl_config(u32 func, u32 data) { return 0; }
+static inline int sip_smc_secure_reg_write(u32 addr_phy, u32 val) { return 0; }
+static inline u32 sip_smc_secure_reg_read(u32 addr_phy) { return 0; }
+
+static inline struct arm_smccc_res sip_smc_bus_config(u32 arg0, u32 arg1, u32 arg2)
+{
+	struct arm_smccc_res tmp = { .a0 = SIP_RET_NOT_SUPPORTED };
+	return tmp;
+}
+
+static inline struct dram_addrmap_info *sip_smc_get_dram_map(void)
+{
+	return NULL;
+}
+
+static inline int sip_smc_access_mem_os_reg(u32 func,
+					    u32 id,
+					    u32 *val)
+{
+	return 0;
+}
+
+static inline int sip_smc_amp_config(u32 sub_func_id,
+				     u32 arg1,
+				     u32 arg2,
+				     u32 arg3)
+{
+	return 0;
+}
+
+static inline struct arm_smccc_res sip_smc_get_amp_info(u32 sub_func_id,
+							u32 arg1)
+{
+	struct arm_smccc_res tmp = { .a0 = SIP_RET_NOT_SUPPORTED, };
+
+	return tmp;
+}
+
+static inline struct arm_smccc_res sip_smc_get_pvtpll_info(u32 sub_func_id,
+							   u32 arg1)
+{
+	struct arm_smccc_res tmp = { .a0 = SIP_RET_NOT_SUPPORTED, };
+
+	return tmp;
+}
+
+static inline struct arm_smccc_res sip_smc_pvtpll_config(u32 sub_func_id,
+							 u32 arg1, u32 arg2,
+							 u32 arg3, u32 arg4,
+							 u32 arg5, u32 arg6)
+{
+	struct arm_smccc_res tmp = { .a0 = SIP_RET_NOT_SUPPORTED, };
+
+	return tmp;
+}
+
+static inline void __iomem *sip_hdcp_request_share_memory(int id)
+{
+	return NULL;
+}
+
+static inline struct arm_smccc_res sip_hdcp_config(u32 arg0, u32 arg1, u32 arg2)
+{
+	struct arm_smccc_res tmp = { .a0 = SIP_RET_NOT_SUPPORTED };
+
+	return tmp;
+}
+
+static inline ulong sip_cpu_logical_map_mpidr(u32 cpu) { return 0; }
+
+/***************************fiq debugger **************************************/
+static inline void sip_fiq_debugger_enable_fiq
+			(bool enable, uint32_t tgt_cpu) { return; }
+
+static inline void sip_fiq_debugger_enable_debug(bool enable) { return; }
+static inline int sip_fiq_debugger_uart_irq_tf_init(u32 irq_id,
+						    sip_fiq_debugger_uart_irq_tf_cb_t callback_fn)
+{
+	return 0;
+}
+
+static inline int sip_fiq_debugger_set_print_port(u32 port_phyaddr,
+						  u32 baudrate)
+{
+	return 0;
+}
+
+static inline int sip_fiq_debugger_request_share_memory(void) { return 0; }
+static inline int sip_fiq_debugger_get_target_cpu(void) { return 0; }
+static inline int sip_fiq_debugger_switch_cpu(u32 cpu) { return 0; }
+static inline int sip_fiq_debugger_sdei_switch_cpu(u32 cur_cpu, u32 target_cpu,
+						   u32 flag) { return 0; }
+static inline int sip_fiq_debugger_is_enabled(void) { return 0; }
+static inline int sip_fiq_debugger_sdei_get_event_id(u32 *fiq, u32 *sw_cpu, u32 *flag)
+{
+	return SIP_RET_NOT_SUPPORTED;
+}
+
+static inline int sip_fiq_control(u32 sub_func, u32 irq, unsigned long data)
+{
+	return 0;
+}
+
+static inline int sip_wdt_config(u32 sub_func,
+				 u32 arg1,
+				 u32 arg2,
+				 u32 arg3)
+{
+	return 0;
+}
+
+static inline int sip_hdmirx_config(u32 sub_func,
+				    u32 arg1,
+				    u32 arg2,
+				    u32 arg3)
+{
+	return SIP_RET_NOT_SUPPORTED;
+}
+
+static inline int sip_hdcpkey_init(u32 hdcp_id)
+{
+	return 0;
+}
+
+static inline int sip_smc_mcu_config(unsigned long mcu_id,
+				     unsigned long func,
+				     unsigned long arg2)
+{
+	return SIP_RET_NOT_SUPPORTED;
+}
+#endif
+
+/* 32-bit OP-TEE context, never change order of members! */
+struct sm_nsec_ctx {
+	u32 usr_sp;
+	u32 usr_lr;
+	u32 irq_spsr;
+	u32 irq_sp;
+	u32 irq_lr;
+	u32 fiq_spsr;
+	u32 fiq_sp;
+	u32 fiq_lr;
+	u32 svc_spsr;
+	u32 svc_sp;
+	u32 svc_lr;
+	u32 abt_spsr;
+	u32 abt_sp;
+	u32 abt_lr;
+	u32 und_spsr;
+	u32 und_sp;
+	u32 und_lr;
+	u32 mon_lr;
+	u32 mon_spsr;
+	u32 r4;
+	u32 r5;
+	u32 r6;
+	u32 r7;
+	u32 r8;
+	u32 r9;
+	u32 r10;
+	u32 r11;
+	u32 r12;
+	u32 r0;
+	u32 r1;
+	u32 r2;
+	u32 r3;
+};
+
+/* 64-bit ATF context, never change order of members! */
+struct gp_regs_ctx {
+	u64 x0;
+	u64 x1;
+	u64 x2;
+	u64 x3;
+	u64 x4;
+	u64 x5;
+	u64 x6;
+	u64 x7;
+	u64 x8;
+	u64 x9;
+	u64 x10;
+	u64 x11;
+	u64 x12;
+	u64 x13;
+	u64 x14;
+	u64 x15;
+	u64 x16;
+	u64 x17;
+	u64 x18;
+	u64 x19;
+	u64 x20;
+	u64 x21;
+	u64 x22;
+	u64 x23;
+	u64 x24;
+	u64 x25;
+	u64 x26;
+	u64 x27;
+	u64 x28;
+	u64 x29;
+	u64 lr;
+	u64 sp_el0;
+	u64 scr_el3;
+	u64 runtime_sp;
+	u64 spsr_el3;
+	u64 elr_el3;
+};
+
+#endif
